File: 0033-search-in-rotated-sorted-array/solution.java
class Solution {
    public int pivot(int[] nums){
        int s=0,e=nums.length-1;
        while(s<e){
            int mid=s+(e-s)/2;
            if(nums[mid]>nums[mid+1]) return mid+1;
            else if(nums[mid]<nums[e]) e=mid;
            else s=mid+1;
        }
        return s;
    }
    public int bsearch(int[] nums,int s,int e,int target){
        while(s<e){
            int mid=s+(e-s)/2;
            if(nums[mid]==target) return mid;
            else if(nums[mid]>target) e=mid;
            else s=mid+1;
        }
        if(nums[s]==target) return s;
        return -1;
    }
    public int search(int[] nums, int target) {
        int p=pivot(nums);
        if(nums[p]==target) return p;
        int s=bsearch(nums,0,p,target);
        if(s!=-1) return s;
        s=bsearch(nums,p,nums.length-1,target);
        return s;
    }
}

File: 0033-search-in-rotated-sorted-array/solution.py
class Solution:
    def peak(self, nums: List[int])->int:
        l=0
        r=len(nums)-1
        while(l<r):
            mid=l+(r-l)//2
            if(nums[mid]<nums[mid-1]):
                return mid
            elif(nums[mid]>nums[mid+1]):
                return mid+1
            elif(nums[mid]<nums[l]):
                r=mid
            else:
                l=mid+1
        return l
    
    def bin_search(self,s,e,target,nums)->int:
        while(s<=e):
            mid=s+(e-s)//2
            if(nums[mid]==target):
                return mid
            elif(nums[mid]>target):
                e=mid-1
            else:
                s=mid+1
        return -1

    def search(self, nums: List[int], target: int) -> int:

        p_ele=self.peak(nums)
        if p_ele==0:
            return self.bin_search(0,len(nums)-1,target,nums)
        if target==nums[p_ele]:
            return p_ele
        elif target>=nums[0]:
            return self.bin_search(0,p_ele-1,target,nums)
        else:
            return self.bin_search(p_ele+1,len(nums)-1,target,nums)
        return -1

    

